src/Learning/MHHistograms/MHDiscretizerMODLHistogram.cpp

```diff
@@ -109,33 +111,50 @@ void MHDiscretizerMODLHistogram::DiscretizeValues(ContinuousVector* cvSourceValu
+		// Le probleme est que l'histogramme interpretable le plus fin est parfois tres complexe, de type "herisson",
+		// ce qui peut derouter un utilisateur non expert. Ce probleme se produit dans environ 20% des cas.
+		// Plutot que de montrer par defaut l'histogramme interpretable le plus fin, on propose d'utiliser la
+		// regle de Terell-Scott  (cf. https://en.wikipedia.org/wiki/Histogram), avec un nombre de bins de (2 N)^1/3
+		// Ce choix permet d'avoir par defaut un histogramme a la fois fin et plus simple a interpreter:
+		// - en evitant des choix trop heuristiques de type "regle du coude", ou nombre maximal de "peaks"
+		// - pas trop parcimonieux (comme la regle de Sturges en log(N))
+		// - ne dependant pas des valeurs (comme la regle de Scott, qui utilise l'ecart type)
+		// - tres simple a calculer
+		// - avec une justification theorique, facile a defendre
+		nTerellScottMaxPartNumber = int(ceil(pow(2 * cvSourceValues->GetSize(), 1.0 / 3)));
+		nMaxPartNumber = nTerellScottMaxPartNumber;
+
+		// La regle de Terell-Scott est pertinente pour eviter les histogrammes de type "herisson" comportant des
+		// centaines, voire des milliers d'intervalles: cela ameliore vraiment le choix de la granularite par defaut.
+		// Par contre, dans le cas des discretisation avec faible nombre d'intervalles, l'application de cette
+		// regle aboutit parfois a un choix de granulatite trop grossiere, avec perte d'information
+		// De facon heuristique, on choisit donc de ne pas appliquer la regle de Terell-Scott s'il y a moins
+		// de 100 intervalles
+		nMaxPartNumber = max(100, nMaxPartNumber);
+
+		// Nombre max partie en prenant en compte l'eventuelle contrainte utilisateur
 		if (nRequiredMaxPartNumber > 0)
+			nMaxPartNumber = min(nRequiredMaxPartNumber, nMaxPartNumber);
+
+		// On recherche l'histogramme interpretable le plus fin qui respecte la contrainte
+		bestHistogram = NULL;
+		for (n = oaResultHistograms.GetSize() - 1; n >= 0; n--)
 		{
-			// On garde le premier histogramme qui respecte la contrainte
-			if (bestHistogram->GetIntervalNumber() > nRequiredMaxPartNumber)
+			histogram = cast(MHHistogram*, oaResultHistograms.GetAt(n));
+
+			// On ne considere que les histogrammes interpretables, c'est a dire tous sauf eventuellement le dernier
+			if (not histogram->GetRaw())
 			{
-				// Parcours des histogramme candidats en partant de l'avant dernier, ce qui permet
-				// de traiter les cas ou le meilleur histogramme etait la version post-optimise ou non
-				bestHistogram = NULL;
-				for (n = oaResultHistograms.GetSize() - 2; n >= 0; n--)
+				if (histogram->GetIntervalNumber() <= nMaxPartNumber)
 				{
-					histogram = cast(MHHistogram*, oaResultHistograms.GetAt(n));
-					if (histogram->GetIntervalNumber() <= nRequiredMaxPartNumber)
-					{
-						bestHistogram = histogram;
-						break;
-					}
+					bestHistogram = histogram;
+					break;
 				}
-				assert(bestHistogram != NULL);
 			}
 		}
+		assert(bestHistogram != NULL);
```

Comment:
How can we know for sure that this `assert` holds? Because `bestHistogram` being non-null depends on the success of at least one test...

Reply:
On le sait d'après la sémantique de oaResultHistograms, qui contient la liste des histogrammes ordonnée par granularité croissante, le premier ne contenant qu'un seul intervalles, nécessaire non `Raw`, et seul le dernier étant potentiellement `Raw`). On doit donc a minimal renvoyer le premier.

Mais dans in contexte non trivial, un petit assert rapide à intégrer permet de le vérifier explicitement en mode debug.
C'est pas cher, ça ne nuit pas, et ça permet de se prémunir d'un changement potentiel de l'implémentation ou d'un effet de bord non anticipé.

Reply:
Ok, then I would just add a comment before the `assert`, like:

```c++
// Au moins le premier histogramme est non-Raw, car il ne contient qu'un seul intervalle
```

But I have another naïve question: does the semantic of the `oaResultsHistogram` guarantee that `.GetSize()` always returns something that is `>= 1`?

Reply:
Oui, on a toujours au moins un histogramme.

Sinon, la politique d'usage des assertions est très pragmatique, il y en a des milliers dans le code.
On évite d'être redondante avec le code ou les commentaires des méthodes utilisées, qu'il faudrait également maintenir sinon, avec une double maintenance couteuse, avec un rapport coût/bénéfice amoindri.

Reply:
Ok.